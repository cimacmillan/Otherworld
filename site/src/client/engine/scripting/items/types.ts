import { GameItem } from "../../../resources/manifests/Items";

export interface Item {
    id: GameItem;
    spriteIcon: string;
    stackable: boolean;
    name: string;
    value?: number;
    behaviours: ItemComponent[];
    category: ItemCategory;
    description: string;
}

export enum ItemCategory {
    CONSUMABLE = "Consumable",
    CRAFTING = "Crafting",
    PRECIOUS = "Precious",
    KEY = "Key",
}

export interface ItemMetadata {
    item: Item;
    count: number;
}

export interface Inventory {
    items: ItemMetadata[];
}

export const getEmptyInventory = (): Inventory => ({
    items: [],
});

export type ItemComponent = ItemHealsPlayer | MakesNoiseWhenConsumed;

export interface ItemHealsPlayer {
    type: ItemComponentType.HEALS_PLAYER;
    amount: number;
}

export interface MakesNoiseWhenConsumed {
    type: ItemComponentType.MAKES_NOISE_WHEN_CONSUMED;
}

export enum ItemComponentType {
    HEALS_PLAYER = "HEALS_PLAYER",
    MAKES_NOISE_WHEN_CONSUMED = "MAKES_NOISE_WHEN_CONSUMED",
}
