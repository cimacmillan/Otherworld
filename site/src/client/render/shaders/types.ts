
interface UniformPositions {
    [key: string]: WebGLUniformLocation;
}

interface AttributePositions {
    [key: string]: number;
}

interface CompiledShader {
    shaderId: WebGLProgram;
    uniform: UniformPositions;
    attribute: AttributePositions;
}
