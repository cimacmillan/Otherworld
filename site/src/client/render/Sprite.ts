import { ScreenBuffer, DepthBuffer } from ".";
import { Camera, Sprite } from "../types";
import { vec_sub, vec_add, vec_rotate, vec_distance, clipToRange, swapSort } from "../util/math";
import { basicShader } from "./Shader";

function drawSprite(screen: ScreenBuffer, depth_buffer: DepthBuffer, camera: Camera, sprite: Sprite) {
    const projectPosition = sprite.projectPosition!;

    const diff = vec_sub(sprite.position, {x: 5, y: 5});
    sprite.position = vec_add(vec_rotate(diff, 0.04 / Math.pow(vec_distance(diff), 2) ), {x: 5, y: 5});

    // Angle of 0 is looking up? 
    const distance = -projectPosition.y;

    if (distance < camera.clip_depth) {
        return;
    }

    const projectionMultiplier = (camera.focal_length / distance) * camera.space_to_pixel;
    
    let width = sprite.size.x * projectionMultiplier; 
    let height = sprite.size.y * projectionMultiplier;
    
    let x = (projectPosition.x * projectionMultiplier) + (0.5 * screen.width);
    let y = (-sprite.height + camera.height) * projectionMultiplier + (0.5 * screen.height);

    let x1 = clipToRange(Math.floor(x - width/2), 0, screen.width-1);
    let x2 = clipToRange(Math.floor(x + width/2), 0, screen.width-1);
    let y1 = clipToRange(Math.floor(y - height/2), 0, screen.height-1);
    let y2 = clipToRange(Math.floor(y + height/2), 0, screen.height-1);
    
    for (let xPixel = x1; xPixel < x2; xPixel++) {
        for (let yPixel = y1; yPixel < y2; yPixel++) {
            screen.putPixelColour(xPixel, yPixel, basicShader(sprite.colour));
            depth_buffer.setDistance(xPixel, yPixel, distance);
        }
    }

}

export function drawSprites(screen: ScreenBuffer, depth_buffer: DepthBuffer, camera: Camera, sprites: Sprite[]) {
    sprites.forEach((sprite) => {
        sprite.projectPosition = vec_rotate(vec_sub(sprite.position, camera.position), -camera.angle)
    })
    
    swapSort(sprites, (spriteA, spriteB) => (spriteA.projectPosition!.y < spriteB.projectPosition!.y));

    sprites.forEach((sprite) => {
        drawSprite(screen, depth_buffer, camera, sprite);
    });
}