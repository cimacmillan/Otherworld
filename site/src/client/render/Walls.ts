import { DepthBuffer, ScreenBuffer } from ".";
import { Camera, Ray, Vector2D, Wall, Colour } from "../types";
import { convert_unit, interpolate, vec_cross, vec_rotate, vec_sub, clipToRange } from "../util/math";
import { shade } from "./Shader";

export function drawWalls(screen: ScreenBuffer, depth_buffer: DepthBuffer, camera: Camera, walls: Wall[], backgroundColour: Colour) {
    // return drawRaycastedWalls(screen, depth_buffer, camera, walls);
    return drawRasterisedWalls(screen, depth_buffer, camera, walls, backgroundColour);
}


function drawRasterisedWalls(screen: ScreenBuffer, depth_buffer: DepthBuffer, camera: Camera, walls: Wall[], backgroundColour: Colour) {
    for (let wallIndex = 0; wallIndex < walls.length; wallIndex++) {
        const wall = walls[wallIndex];
        const texcoord = wall.texcoord;
        const texture = wall.texture;

        const wallA = vec_rotate(vec_sub(wall.p0, camera.position), -camera.angle);
        const wallB = vec_rotate(vec_sub(wall.p1, camera.position), -camera.angle);
        let distanceA = -wallA.y;
        let distanceB = -wallB.y;
        let wallX1 = wallA.x;
        let wallX2 = wallB.x;
        let wallU1 = texcoord.start.x;
        let wallU2 = texcoord.end.x;
        let wallYT1 = -(wall.height0 + wall.offset0 - camera.height);
        let wallYT2 = -(wall.height1 + wall.offset1 - camera.height);
        let wallYB1 = -(wall.offset0 - camera.height);
        let wallYB2 = -(wall.offset1 - camera.height);

        if (distanceA < camera.clip_depth && distanceB < camera.clip_depth ) {
            continue;
        }

        if (distanceA > camera.far_clip_depth && distanceB > camera.far_clip_depth ) {
            continue;
        }

        if (distanceA < camera.clip_depth) {
            const depthAlpha = (camera.clip_depth - distanceA) / (distanceB - distanceA);
            wallX1 = wallX1 + (wallX2 - wallX1) * depthAlpha;
            wallU1 = wallU1 + (wallU2 - wallU1) * depthAlpha;
            wallYT1 = wallYT1 + (wallYT2 - wallYT1) * depthAlpha;
            wallYB1 = wallYB1 + (wallYB2 - wallYB1) * depthAlpha;
            distanceA = camera.clip_depth;
        }

        if (distanceB < camera.clip_depth) {
            const depthAlpha = (camera.clip_depth - distanceB) / (distanceA - distanceB);
            wallX2 = wallX2 + (wallX1 - wallX2) * depthAlpha;
            wallU2 = wallU2 + (wallU1 - wallU2) * depthAlpha;
            wallYT2 = wallYT2 + (wallYT1 - wallYT2) * depthAlpha;
            wallYB2 = wallYB2 + (wallYB1 - wallYB2) * depthAlpha;
            distanceB = camera.clip_depth;
        }

        if (distanceA > camera.far_clip_depth) {
            const depthAlpha = (camera.far_clip_depth - distanceA) / (distanceB - distanceA);
            wallX1 = wallX1 + (wallX2 - wallX1) * depthAlpha;
            wallU1 = wallU1 + (wallU2 - wallU1) * depthAlpha;
            wallYT1 = wallYT1 + (wallYT2 - wallYT1) * depthAlpha;
            wallYB1 = wallYB1 + (wallYB2 - wallYB1) * depthAlpha;
            distanceA = camera.far_clip_depth;
        }

        if (distanceB > camera.far_clip_depth) {
            const depthAlpha = (camera.far_clip_depth - distanceB) / (distanceA - distanceB);
            wallX2 = wallX2 + (wallX1 - wallX2) * depthAlpha;
            wallU2 = wallU2 + (wallU1 - wallU2) * depthAlpha;
            wallYT2 = wallYT2 + (wallYT1 - wallYT2) * depthAlpha;
            wallYB2 = wallYB2 + (wallYB1 - wallYB2) * depthAlpha;
            distanceB = camera.far_clip_depth;
        }

        const wallAX = ((wallX1 * camera.focal_length * screen.height) / distanceA) + (0.5 * screen.width);
        const wallBX = ((wallX2 * camera.focal_length * screen.height) / distanceB) + (0.5 * screen.width);

        const clipX1 = wallAX > screen.width - 1 ? screen.width - 1 : (wallAX < 0 ? 0 : wallAX);
        const clipX2 = wallBX > screen.width - 1 ? screen.width - 1 : (wallBX < 0 ? 0 : wallBX);

        const x1 = clipX1 < clipX2 ? clipX1 : clipX2;
        const x2 = clipX1 > clipX2 ? clipX1 : clipX2;

        const persMultiplyA = (1.0 / distanceA) * camera.focal_length * camera.y_view_window * screen.height;
        const persMultiplyB = (1.0 / distanceB) * camera.focal_length * camera.y_view_window * screen.height;
        const halfHeight = screen.height / 2;

        const upper_pixel_start = wallYT1 * persMultiplyA + halfHeight;
        const upper_pixel_end = wallYT2 * persMultiplyB + halfHeight;
        const lower_pixel_start = wallYB1 * persMultiplyA + halfHeight;
        const lower_pixel_end = wallYB2 * persMultiplyB + halfHeight;

        const zinv_start = 1.0 / distanceA;
        const zinv_end = 1.0 / distanceB;

        const u_start = wallU1 * zinv_start;
        const u_end = wallU2 * zinv_end;

        for (let x = x1; x < x2; x++) {

            const alpha = (x - wallAX) / (wallBX - wallAX);
            const ialpha = 1 - alpha;
            const invDistance = ((ialpha * zinv_start) + (alpha * zinv_end));
            const distance = 1 / invDistance;

            const upper_pixel = (ialpha * upper_pixel_start) + (alpha * upper_pixel_end);
            const lower_pixel = (ialpha * lower_pixel_start) + (alpha * lower_pixel_end);
            
            const upper_pixel_in_range = upper_pixel >= screen.height ? screen.height - 1 : (upper_pixel < 0 ? 0 : upper_pixel);
            const lower_pixel_in_range = upper_pixel >= screen.height ? screen.height - 1 : (lower_pixel < 0 ? 0 : lower_pixel);

            const u = ((ialpha * u_start) + (alpha * u_end)) * distance;

            for (let y = upper_pixel_in_range; y <= lower_pixel_in_range; y++) {

                if (depth_buffer.isCloserAndSetInv(~~x, ~~y, invDistance)) {
                    const beta = (y - upper_pixel) / (lower_pixel - upper_pixel);
                    const v = (texcoord.end.y * beta) + ((1.0 - beta) * texcoord.start.y);

                    const colour = shade(texture, u, v, distance, camera.far_clip_depth);

                    if (colour && colour.a > 0) {     
                        screen.putPixelColour(~~x, ~~y, colour, distance, camera.far_clip_depth, backgroundColour);
                    }
                }
            }
        }
    }
}

