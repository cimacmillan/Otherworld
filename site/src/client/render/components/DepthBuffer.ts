import { fill } from "../../util/math/Array";

export class DepthBuffer {

    data: number[];
    width: number;
    height: number;

    constructor(width: number, height: number) {
        this.data = Array(width * height).fill(0);
        this.width = width;
        this.height = height;
    }

    isCloser(x: number, y: number, distance: number) {
        let index = (y * this.width) + x;
        return 1.0 / distance > this.data[index];
    }

    setDistance(x: number, y: number, distance: number) {
        let index = (y * this.width) + x;
        this.data[index] = 1.0 / distance;
    }

    reset() {
        fill(this.data, 0);
    }
}
