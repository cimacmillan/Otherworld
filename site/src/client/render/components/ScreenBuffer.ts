import { Colour } from "../../types";
import { fillPattern } from "../../util/math/SyncedArray";

export class ScreenBuffer {
  // public image_data: ImageData;
  private gl: WebGLRenderingContext;
  public width: number;
  public height: number;

  constructor(gl: WebGLRenderingContext, width: number, height: number) {
    this.gl = gl;
    this.width = width;
    this.height = height;
  }

  public getOpenGL() {
    return this.gl;
  }

  // public putPixel(x: number, y: number, red: number, green: number, blue: number, alpha: number) {
  //     const pixelindex = (y * this.width + x) * 4;
  //     this.image_data.data[pixelindex] = red;
  //     this.image_data.data[pixelindex + 1] = green;
  //     this.image_data.data[pixelindex + 2] = blue;
  //     this.image_data.data[pixelindex + 3] = alpha;
  // }

  // public putPixelColour(x: number, y: number, colour: Colour, depth: number, farClip: number, backgroundColour: Colour) {

  //     let colourMult = (depth / farClip);
  //     colourMult *= colourMult;

  //     // Breaking the rules but it's faster
  //     this.putPixel(x, y,
  //         ~~(colour.r + colourMult * (backgroundColour.r - colour.r)),
  //         ~~(colour.g + colourMult * (backgroundColour.g - colour.g)),
  //         ~~(colour.b + colourMult * (backgroundColour.b - colour.b)),
  //         ~~(colour.a + colourMult * (backgroundColour.a - colour.a)),
  //         );
  // }

  // public fillBackground(red: number, green: number, blue: number, alpha: number) {
  //     fillPattern(this.image_data.data, [red, green, blue, alpha]);
  // }

  // public reset(colour: Colour) {
  //     this.fillBackground(colour.r, colour.g, colour.b, 255);
  // }
}
