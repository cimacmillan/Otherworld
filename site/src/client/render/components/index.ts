export * from "./CanvasComponent"
export * from "./DepthBuffer"
export * from "./ScreenBuffer"