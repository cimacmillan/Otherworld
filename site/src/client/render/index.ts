export * from "./components/CanvasComponent";
export * from "./components/DepthBuffer";
export * from "./components/ScreenBuffer";
export * from "./Render";
export * from "./Planes";
