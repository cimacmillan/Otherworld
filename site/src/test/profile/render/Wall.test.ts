import { SAMPLE_LIL, profile, SAMPLE_SMA, SAMPLE_MEDIUM_RARE } from "../ProfileUnit";
import { drawSprite, drawSprites} from "../../../client/render/Sprite";
import { ScreenBuffer } from "../../../client/render/components/ScreenBuffer";
import { DepthBuffer } from "../../../client/render/components/DepthBuffer";
import { Camera, Texture, Sprite, Wall } from "../../../client/types";
import { convertToFastTexture } from "../../../client/util/loader/TextureLoader";
import { drawWalls } from "../../../client/render/Walls";

jest.mock("../../../client/render/components/ScreenBuffer");
jest.mock("../../../client/render/components/DepthBuffer");

const mockedScreenbuffer = ScreenBuffer as jest.Mock<ScreenBuffer>;
const mockedDepthBuffer = DepthBuffer as jest.Mock<DepthBuffer>;

const randomTexture = (width: number, height: number): Texture => {
    let data = new Uint8ClampedArray(width * height * 4);
    return {data: {data, width, height}, width, height};
}

const MANY_WALLS = 100;

describe("Wall Profile", () => {
    let screenBuffer: ScreenBuffer;
    let depthBuffer: DepthBuffer;
    let camera: Camera;

    beforeEach(() => {
        mockedScreenbuffer.mockReset();
        screenBuffer = new mockedScreenbuffer();
        depthBuffer = new mockedDepthBuffer();
        screenBuffer.width = 256;
        screenBuffer.height = 256;
        camera = {
            position: { x: 0.0, y: 4.0 }, 
            angle: 0.0, 
            focal_length: 1, 
            height: 0.5, 
            x_view_window: 1, 
            y_view_window: 1, 
            clip_depth: 0.1
        };
    });

    describe("Many Walls", () => {

        let walls: Wall[];
            
        beforeEach(() => {
            walls = [];
            const texture = convertToFastTexture(randomTexture(64, 64));

            for (let i = 0; i < 10; i++) {
                walls.push({ 
                    p0: { x: -5.0, y: 0.0 }, 
                    p1: { x: 5.0, y: 0.0 }, 
                    height0: 1, 
                    height1: 1, 
                    offset0: 0, 
                    offset1: 0,
                    texture,
                    texcoord: {
                        start: {
                            x: 0,
                            y: 0
                        },
                        end: {
                            x: 10,
                            y: 1
                        }
                    }
                });
            }
        });

        test("rendering walls", () => {
            profile(MANY_WALLS, () => drawWalls(screenBuffer, depthBuffer, camera, walls));
        });

    });

});

