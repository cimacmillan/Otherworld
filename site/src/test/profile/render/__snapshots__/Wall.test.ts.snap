// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Wall Profile Many Walls rendering walls 1`] = `4132`;
